### DISCUSSION TITLE
Pattern extraction from string to tuple

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/csharplang#6438] [Language Ideas] [answered]
Is there any mechanism in C# today to match to a pattern in a string and assign the various matching variables to a tuple? 

I frequently find myself parsing out values from strings doing something like:

```
var prefix = "AttachedProperties_";
var id = Guid.NewGuid();
var name = "PropertyAbc"
var iteration = 1;
var key = $"{prefix}__{id}__{name}__{iteration}"; //Key value of "AttachedProperties__f2b52338-8e52-4856-b8c6-70b157bd4521__PropertyAbc__1"
```

Later on, I need to extract out each of these values. Because I delimited everything by double underscores, I can split out the string by "__" and capture each of the values:

```
var key = "AttachedProperties__f2b52338-8e52-4856-b8c6-70b157bd4521__PropertyAbc__1";
var splitVal = key.Split("__");
var prefix = splitVal[0];
var id = splitVal[1];
var name = splitVal[2];
var iteration = splitVal[3];
```

But this increasingly reminds me of the days before we had tuples and we had to create all these one-time objects just to return sets of values. Since I know how the text is delimited, it would be great if there were some mechanism by which I could do something like the following to simply have all the values extracted, then dropped into a tuple for me:

`var (string Prefix, string Id, string Name, string Iteration) = key.ToTuple("__");`

I took a stab at building this out as an extension, but I'm stumped on how I'd return a generic tuple (can't figure out how I'd determine the number of keys in TTuple and it doesn't appear I can constrain the generic type to a tuple).

Even if not using a tuple, is there some more elegant way to do this besides splitting the string and assigning each index to a variable to use downstream?

Thanks!

**Comment by ufcpp** [comment, marked as answer]
No new syntax is required.

```cs
var prefix = "AttachedProperties_";
var id = Guid.NewGuid();
var name = "PropertyAbc";
var iteration = 1;
var key = $"{prefix}__{id}__{name}__{iteration}";

var (Prefix, Id, Name, Iteration) = key.ToTuple("__");

struct SplittedString
{
    private string[] _values;

    public SplittedString(string[] values) => _values = values;
    public SplittedString(string text, char delimiter) : this(text.Split(delimiter)) { }
    public SplittedString(string text, string delimiter) : this(text.Split(delimiter)) { }

    public void Deconstruct(out string arg1) => arg1 = _values[0];
    public void Deconstruct(out string arg1, out string arg2) => (arg1, arg2) = (_values[0], _values[1]);
    public void Deconstruct(out string arg1, out string arg2, out string arg3) => (arg1, arg2, arg3) = (_values[0], _values[1], _values[2]);
    public void Deconstruct(out string arg1, out string arg2, out string arg3, out string arg4) => (arg1, arg2, arg3, arg4) = (_values[0], _values[1], _values[2], _values[3]);
}

static class ParseExtensions
{
    public static SplittedString ToTuple(this string text, string delimiter) => new(text, delimiter);
    public static SplittedString ToTuple(this string text, char delimiter) => new(text, delimiter);
}
```

**Reply by WhitWaldo** [reply]
Thank you very much!
